boot/freeldr/freeldr/arch/drivers/hwide.c

```diff
@@ -2,940 +2,1252 @@
+        TotalSectors = Cylinders * Heads * SectorsPerTrack;
+    }
+
+    if (TotalSectors == 0)
+    {
+        ERR("Unknown geometry\n");
+        return FALSE;
+    }
+
+    DeviceUnit->P.TotalSectors = TotalSectors;
+    DeviceUnit->P.Cylinders = Cylinders;
+    DeviceUnit->P.Heads = Heads;
+    DeviceUnit->P.SectorsPerTrack = SectorsPerTrack;
+    DeviceUnit->P.SectorSize = AtaDevBytesPerLogicalSector(IdentifyData);
+    ASSERT(DeviceUnit->P.SectorSize >= 512);
+    DeviceUnit->P.SectorSize = min(DeviceUnit->P.SectorSize, 512);
+
+    AtapAtaSetMultipleMode(DeviceUnit);
+
+    return TRUE;
 }
 
 static
 BOOLEAN
-WaitForBusy(
-    IN UCHAR Channel,
-    IN ULONG Timeout)
+AtapAnalyzeIdenifyData(
+    _In_ PHW_DEVICE_UNIT DeviceUnit,
+    _In_ ATA_DEVICE_CLASS DeviceClass)
 {
-    ASSERT(Timeout != 0);
+    PIDENTIFY_DEVICE_DATA Id = &DeviceUnit->IdentifyDeviceData;
+    ULONG i;
 
-    while (Timeout--)
+    /* Verify the checksum */
+    if (!AtaDevIsIdentifyDataValid(Id))
     {
-        StallExecutionProcessor(10);
+        ERR("Identify data CRC error\n");
+        return FALSE;
+    }
 
-        if ((AtaReadPort(Channel, IDX_IO1_i_Status) & IDE_STATUS_BUSY) == 0)
-            return TRUE;
+    if (DeviceClass == DEV_ATAPI)
+    {
+        DeviceUnit->P.Flags |= ATA_DEVICE_ATAPI;
+
+        if (Id->ModelNumber[0] == 'N' &&
+            Id->ModelNumber[1] == 'E' &&
+            Id->ModelNumber[2] == 'C' &&
+            Id->ModelNumber[3] == ' ')
+        {
+            DeviceUnit->P.Flags |= ATA_DEVICE_IS_NEC_CDR260;
+        }
     }
-    return FALSE;
+
+    /* Swap byte order of the ASCII data */
+    for (i = 0; i < sizeof(Id->SerialNumber) / 2; ++i)
+        ((PUSHORT)Id->SerialNumber)[i] = RtlUshortByteSwap(((PUSHORT)Id->SerialNumber)[i]);
+
+    for (i = 0; i < sizeof(Id->FirmwareRevision) / 2; ++i)
+        ((PUSHORT)Id->FirmwareRevision)[i] = RtlUshortByteSwap(((PUSHORT)Id->FirmwareRevision)[i]);
+
+    for (i = 0; i < sizeof(Id->ModelNumber) / 2; ++i)
+        ((PUSHORT)Id->ModelNumber)[i] = RtlUshortByteSwap(((PUSHORT)Id->ModelNumber)[i]);
+
+    TRACE("MN  '%.*s'\n", sizeof(Id->ModelNumber), Id->ModelNumber);
+    TRACE("FR  '%.*s'\n", sizeof(Id->FirmwareRevision), Id->FirmwareRevision);
+    TRACE("S/N '%.*s'\n", sizeof(Id->SerialNumber), Id->SerialNumber);
+
+    return TRUE;
 }
 
 static
-VOID
-AtaHardReset(IN UCHAR Channel)
+BOOLEAN
+AtapInitDevice(
+    _In_ PHW_DEVICE_UNIT DeviceUnit,
+    _In_ ATA_DEVICE_CLASS DeviceClass)
 {
-    TRACE("AtaHardReset(Controller %d)\n", Channel);
+    if (!AtapAnalyzeIdenifyData(DeviceUnit, DeviceClass))
+        return FALSE;
 
-    AtaWritePort(Channel, IDX_IO2_o_Control, IDE_DC_RESET_CONTROLLER);
-    StallExecutionProcessor(100000);
-    AtaWritePort(Channel, IDX_IO2_o_Control, IDE_DC_REENABLE_CONTROLLER);
-    StallExecutionProcessor(5);
-    WaitForBusy(Channel, ATA_STATUS_TIMEOUT);
+    if (DeviceClass == DEV_ATAPI)
+        return AtapAtapiInitDevice(DeviceUnit);
```

Comment:
`AtapInitAtapiDevice` (to make it symmetric with the other function)?
Otherwise, see the other alternative suggestion below.